Execution semantics

ngzhian opened the pull request:
Preview at https://www.ngzhian.com/simd/core/exec/instructions.html#simd-instructions.

ngzhian commented:
@binji I'm still working on shuffle, swizzle, load splat and load extends, the rest of the ops are mostly done, definitely still needs cleanup. Hoping you can take a preliminary look.

ngzhian commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,494 @@ Where the underlying operators are non-deterministic, because they may return on
+SIMD Instructions
+~~~~~~~~~~~~~~~~~~~~
+
+SIMD instructions are defined in terms of generic numeric operators applied lane-wise based on the :ref:`shape <syntax-simd-shape>`.
```
2. This definition of simd operations rely on the tweak to numeric operations, that lifts numeric operations to sequences of input.

ngzhian commented on document/core/exec/numerics.rst:
```diff
@@ -38,6 +38,23 @@ When several of these placeholders occur in a single clause, then they must be r
       \fcopysign_N(- p_1, + p_2) &=& + p_1 \\
       \end{array}
 
+Numeric operators are lifted to input sequences by applying the operator element-wise, returning a sequence of results. When there are multiple inputs, they must be of equal length.
```
1. This tweak to numeric operations is useful for making the definition of simd operations slightly easier, since each simd operation. Each simd operation can simply pass a sequence of values to the numeric ops.

ngzhian commented on document/core/exec/numerics.rst:
```diff
@@ -135,6 +152,143 @@ When a number is stored into :ref:`memory <syntax-mem>`, it is converted into a
+
+.. math::
+   \begin{array}{lll@{\qquad}l}
+   \simdto_{t\K{x}N}(c) &=&
```
3. This is the key aux function used for almost all simd operations, it converts a i128 into a sequence of numeric values. This is generalized over all shapes (doesn't care about i32x4 or f32x4, they are both converted to a sequence of length 4).

ngzhian replied:
we already have `shape` defined, it's a non-terminal that represents "i32x4" and friends.
Would calling this function `shape` as well be confusing?
Maybe we can call it `to_shape` or something.

rossberg replied:
Hm, yeah, it might be a bit odd. In math you don't usually have "to" prefixes. How about "shaped", somewhat analogous to "unpacked"?

ngzhian replied:
Yea shaped works, `shaped_{i64x2}(c)`, though I have to think it for half a second to see what it does - the grammar doesn't look like something I'm used to (this is more declarative than verb-y)

Couple more suggestions we might want to consider (in no order of preference)
-  `elements_{i64x2}(c)`
-  `scalarize_{i64x2}(c)`
-  `split_{i64x2}(c)`
-  `sequence_{i64x2}(c)`
-  `lanes_{i64x2}(c)` (this is a collision with an existing definition, which is an auxiliary function that returns the number of lanes, but we can rename that to num_lanes)

ngzhian replied:
@rossberg what do you think about the list of suggestion, does any one look better than the rest?

ngzhian replied:
@rossberg any thoughts? Do you still prefer `shaped`?

rossberg replied:
Ah, sorry, I somehow had missed these. `lanes` actually sounds best to me. How about we rename the existing `lanes` function to `dim`?

ngzhian replied:
Sg, renamed `lanes` to `dim`, then renamed `simdto` to `lanes`, in two separate commit, hopefully easier to review.

ngzhian commented on document/core/exec/numerics.rst:
```diff
@@ -135,6 +152,143 @@ When a number is stored into :ref:`memory <syntax-mem>`, it is converted into a
+   \end{array}
+
+..
+   bits_{i8}^{-1}(b^*[0 \slice 8])~
```
4. This section is all my previous attempts to define the \simdto function, can be ignored. I kept them around for reference and will remove them before merge (if we are happy with the current definition above).

ngzhian commented on document/core/exec/numerics.rst:
```diff
@@ -303,6 +457,18 @@ The integer result of predicates -- i.e., :ref:`tests <syntax-testop>` and :ref:
    it holds that :math:`i_1 = i_2\cdot\idivs(i_1, i_2) + \irems(i_1, i_2)`.
 
 
+.. _op-inot:
```
5. A couple of new numeric ops are defined that are pretty straightforward. We need this as SIMD has some numeric instructions, like or and andnot, that are not yet defined.

ngzhian commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,494 @@ Where the underlying operators are non-deterministic, because they may return on
+
+..
+  .. math::
+     \begin{array}{lll@{\qquad}l}
```
6. This commented section is how I defined simd operations *if* we did not lift numeric operations to work on input sequences. I think it is harder to read, many \dots, and more ambiguous.

ngzhian reviewed (commented):
I left some review notes (numbered based on suggested ordering to look at) to highlight bigger changes or extensions to existing definition.

binji commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,489 @@ Where the underlying operators are non-deterministic, because they may return on
+
+3. Assert: due to :ref:`validation <valid-simd-replace-lane>`, a value of :ref:`value type <syntax-valtype>` :math:`t_1` is on the top of the stack.
+
+4. Pop the value :math:`t_1.\CONST~v_1` from the stack.
```
why `v_1` instead of `c_1`?

ngzhian replied:
I wanted to differentiate them a little more, since `c_1` is always a `v128`, but `v_1` is `i32` or `i64`. I'll change it to `c`, it's more consistent.

binji commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,489 @@ Where the underlying operators are non-deterministic, because they may return on
+
+2. Pop the value :math:`\V128.\CONST~c_2` from the stack.
+
+4. Pop the value :math:`\V128.\CONST~c_1` from the stack.
```
typo `3.`

binji commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,489 @@ Where the underlying operators are non-deterministic, because they may return on
+
+.. _exec-simd-vshiftop:
+
+:math:`t\K{x}N\K{.}\vshiftop`
```
why not `shape` here?

ngzhian replied:
I wanted the `t` and the `N` to specify the subscript to `vshiftop`. `shape` will work here too, then I will need `\lanes(\shape)` and `\packed(\shape)` instead of `t` and `N`.

binji replied:
Yeah, looks fine to me either way, I just thought you were using `shape` below

ngzhian replied:
Yea that was a mistake, i fixed them to use `txN`.

binji commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,489 @@ Where the underlying operators are non-deterministic, because they may return on
+
+4. Let :math:`i_1^n` be the sequence :math:`\simdto_{\shape}(c_1)`
+
+5. Let :math:`i` be the result of computing :math:`\bool((i_1 != 0)^n)`.
```
use `\neq` instead of `!=`

binji replied:
Also I'm not sure how to understand the logic here -- we convert each value in the sequence to a true or false value, then apply bool to convert to a sequence of 0 and 1, right? When is it collapsed to a single integer?

ngzhian replied:
Ah I'm missing something in the definition of `bool` to say that it works on a sequence on integers as well (AND over all of them).

binji commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,489 @@ Where the underlying operators are non-deterministic, because they may return on
+
+2. Pop the value :math:`\V128.\CONST~c_1` from the stack.
+
+3. Let :math:`i` be the result of computing :math:`\ine_{\K{128}}(c_1, 0)`.
```
missing closing `)`

binji commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,489 @@ Where the underlying operators are non-deterministic, because they may return on
+
+4. Let :math:`B` be the :ref:`bit width <syntax-valtype>` :math:`|t|` of :ref:`value type <syntax-valtype>` :math:`t`.
+
+5. Let :math:`i_2^N` sequence as a result of computing :math:`\ilts_{B}(i_1^N, 0^N)`.
```
s/sequence/be the sequence/

binji commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,489 @@ Where the underlying operators are non-deterministic, because they may return on
+
+3. Pop the value :math:`t.\CONST~c_1` from the stack.
+
+4. Let :math:`n` be the integer :math:`\lanes(\shape)`.
```
below `N` is often used for the lane count, maybe use that instead for consistency?

binji commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,489 @@ Where the underlying operators are non-deterministic, because they may return on
+
+5. Let :math:`i_2^N` sequence as a result of computing :math:`\ilts_{B}(i_1^N, 0^N)`.
+
+6. Let :math:`c` be the integer :math:`ibits_{32}^{-1}(i_2^N)`.
```
link to ibits

binji replied:
This feels a bit odd since N < 32, maybe pad with 0 bits?

binji commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,489 @@ Where the underlying operators are non-deterministic, because they may return on
+
+4. Let :math:`c` be the result of computing :math:`\simdof_{t_2\K{x}N}((\extend^{\sx}_{t_1,t_2}(i_1))^N)`
+
+6. Push the value :math:`\V128.\CONST~c` onto the stack.
```
typo `5.`

binji commented on document/core/exec/numerics.rst:
```diff
@@ -135,6 +152,143 @@ When a number is stored into :ref:`memory <syntax-mem>`, it is converted into a
+     c_0~\dots~c_N \\
+     && (\where
+       w = |t| / 8 \wedge
+       b^* = bytes_{\i128}(i) \\ &&
```
clearer if you move this to the next line

binji replied:
use `c` instead of `i`?

binji replied:
link to bytes auxiliary function

binji commented on document/core/exec/numerics.rst:
```diff
@@ -626,6 +804,214 @@ The integer result of predicates -- i.e., :ref:`tests <syntax-testop>` and :ref:
+:math:`\ibitselect_N(i_1, i_2, c)`
+..................................
+
+* Use the bits in :math:`c` to select corresponding bit from :math:`i_1` when 1 and :math:`i_2` when 0.
```
`:math:` around `1` and `0`

binji commented on document/core/exec/numerics.rst:
```diff
@@ -626,6 +804,214 @@ The integer result of predicates -- i.e., :ref:`tests <syntax-testop>` and :ref:
+
+* Let :math:`j` be the :ref:`signed interpretation <aux-signed>` of :math:`i`.
+
+* If :math:`j` greater than or equal to :math:`0`.
```
missing ", then return `i`"

binji commented on document/core/exec/numerics.rst:
```diff
@@ -626,6 +804,214 @@ The integer result of predicates -- i.e., :ref:`tests <syntax-testop>` and :ref:
+:math:`\ibitselect_N(i_1, i_2, c)`
+..................................
+
+* Use the bits in :math:`c` to select corresponding bit from :math:`i_1` when 1 and :math:`i_2` when 0.
```
Maybe rephrase this? I can understand not wanting to write out the exact behavior, but I think it should still match the "Return ..." structure.

ngzhian replied:
Yea agree that this should be rephrased, tried something, but it still looks awkward, especially since the implementation below doesn't match the description.

binji commented on document/core/exec/numerics.rst:
```diff
@@ -626,6 +804,214 @@ The integer result of predicates -- i.e., :ref:`tests <syntax-testop>` and :ref:
+
+.. math::
+   \begin{array}{@{}lcll}
+   \ineg_N(0) &=& 0 \\
```
missing other cases

binji commented on document/core/exec/numerics.rst:
```diff
@@ -626,6 +804,214 @@ The integer result of predicates -- i.e., :ref:`tests <syntax-testop>` and :ref:
+:math:`\ineg_N(i)`
+..................
+
+* Let :math:`j` be the :ref:`signed interpretation <aux-signed>` of :math:`i`.
```
Maybe easiest to define in terms of `x = ~y + 1`, e.g. `inot` and `iadd`

binji commented on document/core/exec/numerics.rst:
```diff
@@ -626,6 +804,214 @@ The integer result of predicates -- i.e., :ref:`tests <syntax-testop>` and :ref:
+   \end{array}
+
+
+.. _op-iaddsat_u:
```
should be `add_sat_`, right?

binji commented on document/core/exec/numerics.rst:
```diff
@@ -626,6 +804,214 @@ The integer result of predicates -- i.e., :ref:`tests <syntax-testop>` and :ref:
+.. math::
+   \begin{array}{lll@{\qquad}l}
+   \iaddsats_N(i_1, i_2) &=& -2^{N-1} & (\iff \signed(i_i) + \signed(i_2) < -2^{N-1})\\
+   \iaddsats_N(i_1, i_2) &=& 2^N-1 & (\iff \signed(i_i) + \signed(i_2) > 2^N-1)\\
```
use `saturate_s` below?

binji commented:
Did a pass on it, I really like extending the numerics, makes it much nicer to read IMO

ngzhian commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,489 @@ Where the underlying operators are non-deterministic, because they may return on
+.. math::
+   \begin{array}{lll@{\qquad}l}
+   \X{op}_{t\K{x}N}(n_1,\dots,n_k) &=&
+     \simdof_{t\K{x}N}(op_t(\simdto_{t\K{x}N}(n_1) ~\dots~ \simdto_{t\K{x}N}(n_k))
```
Just updated.

rossberg reviewed (commented):
Some preliminary comments, haven't read all of exec/instructions yet.

ngzhian commented on document/core/exec/numerics.rst:
```diff
@@ -173,6 +327,13 @@ The integer result of predicates -- i.e., :ref:`tests <syntax-testop>` and :ref:
+
+.. math::
+   \begin{array}{lll@{\qquad}l}
+   \bool(C_1~C_2^*) &=& \bool(C_1) \wedge \bool(C_2^*)
```
Do we need to define how `\bigwedge` works?

rossberg replied:
We could call it out as more auxiliary notation in syntax/conventions, like so:
```
* :math:`\bigwedge s^\ast` and :math:`\bigvee s^\ast` denote the conjunction and disjunction of a sequence of conditions, respectively; if the sequence is empty, then the result is vacuously true, respectively false.
```
OTOH, it's more or less established math notation and fairly intuitive, so I suppose it would be understandable without extra explanation. WDYT?

ngzhian replied:
I agree that it's established notation and fairly intuitive, especially for someone reading formal notation. I'll use `\bigwedge` directly, thanks!

ngzhian commented on document/core/exec/numerics.rst:
```diff
@@ -626,6 +811,195 @@ The integer result of predicates -- i.e., :ref:`tests <syntax-testop>` and :ref:
+
+.. math::
+   \begin{array}{@{}lcll}
+   \ibitselect_N(i_1, i_2, N) &=& \ior_N(\iand_N(i_1, c), \iand_N(i_2, \inot_N(c)))
```
Make sense, I chose to fix the prose to match the current formal definition. And fixed the N too.

ngzhian commented on document/core/exec/numerics.rst:
```diff
@@ -1621,3 +1995,74 @@ Conversions
+
+.. _op-saturate_u:
+
+:math:`\saturateu_N(i)`
```
Yup, fixed.

ngzhian commented on document/core/exec/numerics.rst:
```diff
@@ -135,6 +152,143 @@ When a number is stored into :ref:`memory <syntax-mem>`, it is converted into a
+       (\iff \ibits_{128}(i) = d_1^{64}~d_0^{64}) \\
+     \end{array}
+
+These functions are bijections, so they are invertible. For convenience, we give a name to the inverse function:
```
Sg, removed \simdof and also changed all usages to use the inverse.

binji commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,540 @@ Where the underlying operators are non-deterministic, because they may return on
+
+5. Let :math:`i^{16}` be the sequence :math:`\simdto_{i8x16}(c_1)`.
+
+6. Let :math:`v^{255}` be the concatenation of the two sequences :math:`i^{16}~0^{239}`
```
should be `256` and `240`, right?

ngzhian replied:
Thank you! off-by-one, confused length and indices.

binji commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,540 @@ Where the underlying operators are non-deterministic, because they may return on
+
+6. Let :math:`v^{255}` be the concatenation of the two sequences :math:`i^{16}~0^{239}`
+
+7. Let :math:`d_i` be the result of :math:`d_i = v^{255}[ s^{16}[i] ]`.
```
Can this be `d^16`, and then it doesn't need to be expanded explicitly below?

ngzhian replied:
Yea good idea, reworded it using `\dots`, that folds that sequence in nicely.

binji commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,540 @@ Where the underlying operators are non-deterministic, because they may return on
+
+1. Assert: due to :ref:`validation <valid-simd-shuffle>`, two values of :ref:`value type <syntax-valtype>` |V128| are on the top of the stack.
+
+1. Assert: due to :ref:`validation <valid-simd-shuffle>`, that all :math:`\laneidx_i < 32`.
```
fix step numbers

binji commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,540 @@ Where the underlying operators are non-deterministic, because they may return on
+
+3. Let :math:`i_3^{32}` be the concatenation of the two sequences :math:`i_1^{16}~i_2^{16}`.
+
+4. Let :math:`d_i` be :math:`i_3^{32}[\laneidx_i]`.
```
not sure the convention, but it seems like we'd want `d^16` and `laneidx^16` here too

ngzhian replied:
Reworded it like swizzle, so removed the need for `d_i` here.

binji commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,540 @@ Where the underlying operators are non-deterministic, because they may return on
+
+4. Let :math:`c` be the result of computing :math:`\simdto^{-1}_{t_2\K{x}N}((\extend^{\sx}_{t_1,t_2}(i_1))^N)`
+
+6. Push the value :math:`\V128.\CONST~c` onto the stack.
```
typo: 5.

binji commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,540 @@ Where the underlying operators are non-deterministic, because they may return on
+
+5. Let :math:`i_1^N` be the sequence :math:`\simdto_{t\K{x}N}(c_1)`.
+
+7. Let :math:`c` be :math:`\simdto^{-1}_{t\K{x}N}(\vshiftop_{t}(i_1^N, s^N))`.
```
renumber

binji commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,540 @@ Where the underlying operators are non-deterministic, because they may return on
+
+2. Pop the value :math:`\V128.\CONST~c_1` from the stack.
+
+3. Let :math:`i_1^N` be the sequence :math:`\simdto_{t_1\K{x}M}(c_1)[0 \slice N]`.
```
Maybe just me, but these steps make me uncomfortable unless I know that N = M * 2. Maybe we can assert here? (same for the load extend below).

ngzhian replied:
I'm not sure how to assert here, because we always assert due to validation, and this isn't really a validation as N = M * 2 is guaranteed by the accepted syntax.

binji commented on document/core/exec/instructions.rst:
```diff
@@ -464,6 +998,98 @@ Memory Instructions
+
+    c. Let :math:`c` be the result of computing :math:`\simdto^{-1}_{\iN\K{x}L}(n^L)`.
+
+12. Else:
```
typo: 13.

binji commented on document/core/exec/instructions.rst:
```diff
@@ -464,6 +998,98 @@ Memory Instructions
+
+   a. Let :math:`N` be :math:`M * L`.
+
+10. If :math:`\X{ea} + N/8` is larger than the length of :math:`\X{mem}.\MIDATA`, then:
```
This means that, even though the load-extend instructions read 8 bytes, they'll fail if they can't read 16 bytes. Is that intentional?

EDIT: Ah, n/m. This works because N for extend instructions is actually half the full size.

ngzhian replied:
Yea sorry for the confusion, I took your suggestion to split load extend and load splat up, that should make it clearer!

binji commented on document/core/exec/instructions.rst:
```diff
@@ -464,6 +998,98 @@ Memory Instructions
+.. _exec-load-extend:
+.. _exec-load-splat:
+
+:math:`\V128\K{.}\LOAD{M}\K{x}L\_\sx~\memarg` and :math:`\V128\K{.}\LOAD{N}\K{\_splat}~\memarg`
```
It feels like these should be split into separate definitions, IMO

ngzhian replied:
Yea good point, I'll split them up.

binji commented on document/core/exec/instructions.rst:
```diff
@@ -464,6 +998,98 @@ Memory Instructions
+
+12. Else:
+
+    a. Let :math:`m_i` be the integer for which :math:`\bytes_{\iM}(m_i) = b^\ast[i*M/8 \slice M/8]`.
```
Why can't this just be `b*`?

ngzhian replied:
`b*` is the entire 8 bytes that is loaded, we want to extend it lane-by-lane (`m_i`), hence only taking a slice of size `M/8`.

binji replied:
Ah, that makes sense. Thanks!

binji commented:
It's a little tricky to review this, since there are a lot of changes and the force-pushes mean I have to look at everything. Do you have an overview of the changes since the last time?

ngzhian commented:
Yup I think you can start looking at this commit onwards https://github.com/WebAssembly/simd/pull/340/commits/2725a788440a0ac419dffb7e42ef66d35294bf37 which addresses your review comments previously, the notable changes are:

- swizzle and shuffle got reworded
- fixes in sequence lengths in swizzle
- bullet point numbering fixes
- split out load splat and load extend
- reworded extract lane
- reword simd_to description

And following Andreas' comments:

- rename lanes to dim
- rename simd_to to lanes

binji reviewed (approved):
Sorry for the delay -- didn't recheck everything, but what I did looks good.

ngzhian commented on document/core/exec/instructions.rst:
```diff
@@ -464,6 +985,126 @@ Memory Instructions
+
+11. Let :math:`b^\ast` be the byte sequence :math:`\X{mem}.\MIDATA[\X{ea} \slice N/8]`.
+
+12. Let :math:`m_i` be the integer for which :math:`\bytes_{\iM}(m_i) = b^\ast[i*M/8 \slice M/8]`.
```
Variable shadowing here. The `i` in `m_i` is used as an index to refer to each lane, since each lane is loaded then extended. It has nothing to do with the `i` above which is the memory index.
I changed the `i` in `m_i` to `k`.

ngzhian commented on document/core/exec/instructions.rst:
```diff
@@ -464,6 +985,126 @@ Memory Instructions
+     \begin{array}[t]{@{}r@{~}l@{}}
+     (\iff & \X{ea} = i + \memarg.\OFFSET \\
+     \wedge & \X{ea} + (M*L) \leq |S.\SMEMS[F.\AMODULE.\MIMEMS[0]].\MIDATA| \\
+     \wedge & \bytes_{\iM}(m_i) = S.\SMEMS[F.\AMODULE.\MIMEMS[0]].\MIDATA[\X{ea} + i * (M/8) \slice M/8]) \\
```
Fixed in the same way.

ngzhian commented:
@rossberg can you please take another look at this? All of the outstanding comments have been resolved.

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,527 @@ Where the underlying operators are non-deterministic, because they may return on
+.. math::
+   \begin{array}{lcl@{\qquad}l}
+   (\V128\K{.}\VCONST~c_1)~\shape\K{.}\BITMASK &\stepto& (\I32\K{.}\CONST~c)
+     & (\iff c = \ibits_{32}^{-1}(\ilts_{B}(\lanes_{t\K{x}N}(c), 0^N)))
```
```suggestion
     & (\iff c = \ibits_{32}^{-1}(\ilts_{|t|}(\lanes_{t\K{x}N}(c), 0^N)))
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,527 @@ Where the underlying operators are non-deterministic, because they may return on
+
+.. math::
+   \begin{array}{lcl@{\qquad}l}
+   (\V128\K{.}\VCONST~c_1)~\shape\K{.}\BITMASK &\stepto& (\I32\K{.}\CONST~c)
```
```suggestion
   (\V128\K{.}\VCONST~c_1)~t\K{x}N\K{.}\BITMASK &\stepto& (\I32\K{.}\CONST~c)
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,527 @@ Where the underlying operators are non-deterministic, because they may return on
+
+.. _exec-simd-narrow:
+
+:math:`t_2\K{x}N\K{.}\NARROW\_t_1\K{x}M\_\sx`
```
```suggestion
:math:`t_2\K{x}N\K{.}\NARROW\K{\_}t_1\K{x}M\K{\_}\sx`
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,527 @@ Where the underlying operators are non-deterministic, because they may return on
+
+5. Let :math:`d_1^M` be the result of computing :math:`\narrow^{\sx}_{t_1, t_2}(\lanes_{t_1\K{x}M}(c_1))`.
+
+6. Let :math:`c` be the result of :math:`\lanes^{-1}_{t_2\K{x}N}(d_1^M~d_2^M)`.
```
Does this need an assertion that N = 2M, due to syntactic constraints?

Btw, I'm now confused about the inclusion of bare `\NARROW` in vbinop in the syntax conventions. I suppose that should be `\NARROW_\shape`? Similar question about `\WIDEN`.

ngzhian replied:
> Does this need an assertion that N = 2M, due to syntactic constraints?

If we only get here from valid instructions, then we don't need such an assertion. The valid instructions are already defined such that N = 2M.

> Btw, I'm now confused about the inclusion of bare \NARROW in vbinop in the syntax conventions.

What I intended is that these narrow instructions are validated the same as the other vbinop, but for execution, it "pattern-matches" to this more specific definition of narrow (`shape.vbinop` wouldn't work for narrow/widen). I'm not sure if this works though.

I guess narrow and widen probably shouldn't be included in the syntax conventions, and should be separated out in the validation?

rossberg replied:
> If we only get here from valid instructions, then we don't need such an assertion. The valid instructions are already defined such that N = 2M.

Yup, that's why I suggested saying "due to syntactic constraints". The nature of an assertion of course is that it's always true. But the constraint is not apparent from the pattern match above, so local to this rule, it is not clear otherwise.

> > Btw, I'm now confused about the inclusion of bare \NARROW in vbinop in the syntax conventions.
> 
> What I intended is that these narrow instructions are validated the same as the other vbinop, but for execution, it "pattern-matches" to this more specific definition of narrow (`shape.vbinop` wouldn't work for narrow/widen). I'm not sure if this works though.

You can still do so, but you have to list them in vunop/vbinop with their suffixes, i.e. `\NARROW\K{\_}\shape\K{\_}\sx` and `\WIDEN\K{\_low\_}\shape\K{\_}\sx | \WIDEN\K{\_high\_}\shape\K{\_}\sx`, respectively. The same for `\VTRUNC` and `\VCONVERT`, now that I notice it.

ngzhian replied:
> Yup, that's why I suggested saying "due to syntactic constraints".

Added this assertion.

> You can still do so, but you have to list them in vunop/vbinop with their suffixes,

Got it, fixed.

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -464,6 +985,126 @@ Memory Instructions
 
+.. _exec-load-extend:
+
+:math:`\V128\K{.}\LOAD{M}\K{x}L\_\sx~\memarg`
```
Nit: can we use N instead of L here?

ngzhian replied:
`N` is already used later on, defined to be `M * L`. This follows the other memory instructions which use `N` as the number of bits actually loaded/stored.

rossberg replied:
Yes, but N is an intermediate that does not actually occur in the formal rule, so it's a bit odd skipping it from that perspective.

ngzhian replied:
I see what you mean. I think we could simplify this, since `M*L` should always be 64, we can assert this (due to syntax) early on, and propagate 64 everywhere we use `N` or `M*L`.

rossberg replied:
I would suggest simply inlining M\cdot L -- it is short enough and occurs only in two places I think.

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -464,6 +985,126 @@ Memory Instructions
+   \\ \qquad
+     \begin{array}[t]{@{}r@{~}l@{}}
+     (\iff & \X{ea} = i + \memarg.\OFFSET \\
+     \wedge & \X{ea} + (M*L) \leq |S.\SMEMS[F.\AMODULE.\MIMEMS[0]].\MIDATA| \\
```
```suggestion
     \wedge & \X{ea} + M \cdot L / 8 \leq |S.\SMEMS[F.\AMODULE.\MIMEMS[0]].\MIDATA| \\
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -464,6 +985,126 @@ Memory Instructions
+     (\iff & \X{ea} = i + \memarg.\OFFSET \\
+     \wedge & \X{ea} + (M*L) \leq |S.\SMEMS[F.\AMODULE.\MIMEMS[0]].\MIDATA| \\
+     \wedge & \bytes_{\iM}(m_k) = S.\SMEMS[F.\AMODULE.\MIMEMS[0]].\MIDATA[\X{ea} + k * (M/8) \slice M/8]) \\
+     \wedge & J = M*2 \\
+     \wedge & c = \lanes^{-1}_{\X{i}J\K{x}L}(\extend^{\sx}_{M,J}(m_0) \dots \extend^{\sx}_{M,J}(m_{L-1}))
```
```suggestion
     \wedge & c = \lanes^{-1}_{\X{i}J\K{x}L}(\extend^{\sx}_{M,M\cdot2}(m_0) \dots \extend^{\sx}_{M,M\cdot2}(m_{L-1}))
```

rossberg reviewed (commented):
Sorry, I missed this somehow! Various nits and a number of small bugs. I ran out of steam for the load/store instructions, so I may have to take another pass on those.

ngzhian commented on document/core/exec/numerics.rst:
```diff
@@ -625,6 +696,197 @@ The integer result of predicates -- i.e., :ref:`tests <syntax-testop>` and :ref:
+.. math::
+   \begin{array}{@{}lcll}
+   \iabs_N(i) &=& i & (\iff \signed_N(i) \ge 0) \\
+   \iabs_N(i) &=& \signed_N^{-1}(-j) \mod 2^N & (\iff j = \signed_N(i) \wedge j < 0) \\
```
I think the new change should be:
`\iabs_N(i) &=& -\signed_N(i) \mod 2^N & (\otherwise) \\`

from `\signed_N(j)` to \signed_N(i)`, right?

rossberg replied:
Ah, yes, thanks.

ngzhian commented on document/core/exec/numerics.rst:
```diff
@@ -1620,3 +1882,70 @@ Conversions
+:math:`\sats_N(i)`
+.......................
+
+* Let :math:`j` be the :ref:`signed interpretation <aux-signed>` of :math:`i` of size :math:`N`.
```
This function is not always called with signed integers:
- `\isubsatu_N(i_1, i_2)` calls it with the result of `i_1 - i_2`.
- `\addsatu_N(i_1, i_2)` calls it with the result of `i_1 + i_2`.

I can change these two calls so that we can drop the reinterpretation step here.

ngzhian replied:
Oh oops, this is sat_s, i was looking at sat_u, ignore my comment.

ngzhian commented on document/core/exec/numerics.rst:
```diff
@@ -1620,3 +1882,70 @@ Conversions
+.. math::
+   \begin{array}{lll@{\qquad}l}
+   \satu_N(i) &=& 2^N-1 & (\iff i > 2^N-1)\\
+   \satu_N(i) &=& i & (\otherwise)
```
I think to be consistent, I will change `subsat_u` to pass the arg as a signed interpretation, so everything is consistent (both `satu` and `sats` will both take integers as signed interpretation)

ngzhian replied:
Then in that case we don't need a case for values that are negative in two's complement.

ngzhian replied:
Nvm, we need a case for i < 0, since it is signed.

rossberg replied:
Yes, the arguments need to be allowed to be negative as well as larger than the N value range, so they cannot be Wasm values.

ngzhian commented on document/core/exec/numerics.rst:
```diff
@@ -1620,3 +1882,70 @@ Conversions
+:math:`\narrowu_{M,N}(i)`
+.........................
+
+* Let :math:`j` be the :ref:`signed interpretation <aux-signed>` of :math:`i` of size :math:`M`.
```
Yea the definition of narrow (at least for the SIMD instruction) is that the value is interpreted as a signed value. See https://github.com/WebAssembly/simd/blob/master/proposals/simd/SIMD.md#integer-to-integer-narrowing "Regardless of the whether the operation is signed or unsigned, the input lanes are interpreted as signed integers."

rossberg replied:
Cool, thanks.

ngzhian commented on document/core/exec/numerics.rst:
```diff
@@ -1620,3 +1882,70 @@ Conversions
+
+.. _op-narrow_s:
+
+:math:`\narrows_{M,N}(i)`
```
It looks like this convention of `i<operation>` is for the numeric operators. For these conversion, `extend` is just `extend`, not `iextend`. Should we keep the name as `narrow`?

rossberg replied:
Sunds plausible, but if this has the same status as plain `extend`, then don't you need an `inarrow` wrapper as well, same as for extend? Otherwise, how is it picked up by the operator conventions?

ngzhian replied:
In [exec/instructions.rst](https://www.ngzhian.com/simd/core/exec/instructions.html#t-2-mathsf-x-n-mathsf-xref-syntax-instructions-syntax-instr-simd-mathsf-narrow-t-1-mathsf-x-m-xref-syntax-instructions-syntax-sx-mathit-sx) we don't use the operator conventions, we directly refer to `narrow` with the correct types from the instruction's name.

rossberg replied:
Ah, I see. But that causes another issue that I missed before (sorry): there is a generic execution rule for shape.vbinop, and by the definition of vbinop, that should also apply to narrow. So there are two overlapping exec rules matching narrow. Similarly for vunop and widen.

So unlike what I said before, I think you'll need to remove those instructions from vunop and vbinop -- they don't fit the pattern, because unlike the others, they do not work lane-wise. (And consequently, they need separate typing rules.)

ngzhian replied:
Done! See a5165cb, I created a new production `\vwiden` that just groups `widen_low` and `widen_high`, that way I don't have to repeat this in the validation.
`narrow` also has its own validation rule now.

ngzhian commented on document/core/exec/numerics.rst:
```diff
@@ -1620,3 +1882,70 @@ Conversions
    \end{array}
+
+
+.. _op-sat_u:
```
Done, moved the aux-sat_{,s} up to the top. Ptal.

rossberg commented on document/core/exec/numerics.rst:
```diff
@@ -79,6 +79,30 @@ Conventions:
+       \satu_N(i) &=& i & (\otherwise) \\
+       \end{array}
+
+  * Signed saturation, :math:`\sats_N(i)` clamps :math:`i` to between :math:`-2^{N-1}` and :math:`2^N-1`:
```
```suggestion
  * Signed saturation, :math:`\sats_N(i)` clamps :math:`i` to between :math:`-2^{N-1}` and :math:`2^{N-1}-1`:
```

rossberg commented on document/core/exec/numerics.rst:
```diff
@@ -79,6 +79,30 @@ Conventions:
+  .. math::
+     \begin{array}{lll@{\qquad}l}
+     \sats_N(i) &=& \signed_N^{-1}(-2^{N-1}) & (\iff i < -2^{N-1})\\
+     \sats_N(i) &=& \signed_N^{-1}(2^N-1) & (\iff i > 2^N-1)\\
```
```suggestion
     \sats_N(i) &=& \signed_N^{-1}(2^{N-1}-1) & (\iff i > 2^{N-1}-1)\\
```

rossberg commented on document/core/exec/numerics.rst:
```diff
@@ -809,11 +833,11 @@ The integer result of predicates -- i.e., :ref:`tests <syntax-testop>` and :ref:
 * Let :math:`i` be the result of adding :math:`i_1` and :math:`i_2`.
 
-* Return :math:`\satu_N(i)`.
+* Return :math:`\satu_N(\signed_N(i))`.
```
```suggestion
* Return :math:`\satu_N(i)`.
```
This isn't correct here, since the result of i1+i2 is a plain number, not a two's complement representation.

rossberg commented on document/core/exec/numerics.rst:
```diff
@@ -809,11 +833,11 @@ The integer result of predicates -- i.e., :ref:`tests <syntax-testop>` and :ref:
 .. math::
    \begin{array}{lll@{\qquad}l}
-   \iaddsatu_N(i_1, i_2) &=& \satu_N(i_1 + i_2)
+   \iaddsatu_N(i_1, i_2) &=& \satu_N(\signed_N(i_1 + i_2))
```
```suggestion
   \iaddsatu_N(i_1, i_2) &=& \satu_N(i_1 + i_2)
```

rossberg commented on document/core/exec/numerics.rst:
```diff
@@ -843,11 +867,11 @@ The integer result of predicates -- i.e., :ref:`tests <syntax-testop>` and :ref:
 * Let :math:`i` be the result of subtracting :math:`i_2` from :math:`i_1`.
 
-* Return :math:`\satu_N(i)`.
+* Return :math:`\satu_N(\signed_N(i))`.
```
```suggestion
* Return :math:`\satu_N(i)`.
```

rossberg commented on document/core/exec/numerics.rst:
```diff
@@ -843,11 +867,11 @@ The integer result of predicates -- i.e., :ref:`tests <syntax-testop>` and :ref:
 .. math::
    \begin{array}{lll@{\qquad}l}
-   \isubsatu_N(i_1, i_2) &=& \satu_N(i_1 - i_2)
+   \isubsatu_N(i_1, i_2) &=& \satu_N(\signed_N(i_1 - i_2))
```
```suggestion
   \isubsatu_N(i_1, i_2) &=& \satu_N(i_1 - i_2)
```

ngzhian commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,527 @@ Where the underlying operators are non-deterministic, because they may return on
+   \end{array}
+   \\ \qquad
+     \begin{array}[t]{@{}r@{~}l@{}}
+     (\iff & n = \dim(\shape) \\
+     \wedge & i_1^n = \lanes_{\shape}(c) \\
+     \wedge & i = \bigwedge(i_1 \neq 0)^n
```
I used `\ast` instead of `*`.

ngzhian commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,527 @@ Where the underlying operators are non-deterministic, because they may return on
+.. _exec-simd-extract-lane:
+.. _exec-simd-extract-lane-sx:
+
+:math:`t_1\K{x}N\K{.}\EXTRACTLANE\K{\_}\sx^?~\laneidx`
```
> Nit: rename \laneidx to just x.

Done.

> Also, for stylistic consistency

I used `\shape` in places where we don't need to know the underlying shape. But for instructions like `shift`, `narrow`, etc, it is useful to know the components of the shape, and where `unpacked` wouldn't work (we want to know that it is an `i8` and not a `i32`). I guess we could create another helper for this?

rossberg replied:
I see. You wouldn't need a helper, you could just pattern match later saying something like "Let txN be the shape \shape." But maybe that's causing more harm than benefit, so I'm fine leaving it as is.

ngzhian commented:
@rossberg Thanks for the careful review. I have resolved all the minor issues. Left a couple of responses and questions for you. And apologies for not doing the recent changes at one go. I missed that there was a "Show X hidden conversations" button that had a couple of your review notes, and so I didn't fix those issues during my first go.

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -293,65 +293,64 @@ SIMD instructions are defined in terms of generic numeric operators applied lane
 
-:math:`\K{i8x16.}\SHUFFLE~\laneidx^{16}`
+:math:`\K{i8x16.}\SHUFFLE~x^\ast`
 ........................................
```
Nit: adjust underline

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -293,65 +293,64 @@ SIMD instructions are defined in terms of generic numeric operators applied lane
 1. Assert: due to :ref:`validation <valid-simd-shuffle>`, two values of :ref:`value type <syntax-valtype>` |V128| are on the top of the stack.
 
-2. Assert: due to :ref:`validation <valid-simd-shuffle>`, that all :math:`\laneidx_i < 32`.
+2. Assert: due to :ref:`validation <valid-simd-shuffle>`, that all :math:`x^\ast[i] < 32`.
```
```suggestion
2. Assert: due to :ref:`validation <valid-simd-shuffle>`, for all :math:`x_i` in :math:`x^\ast` it holds that :math:`x_i < 32`.
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -293,65 +293,64 @@ SIMD instructions are defined in terms of generic numeric operators applied lane
+6. Let :math:`i_1^\ast` be the sequence :math:`\lanes_{i8x16}(c_1)`.
 
-7. Let :math:`i_3^{32}` be the concatenation of the two sequences :math:`i_1^{16}~i_2^{16}`.
+7. Let :math:`i` be the concatenation of the two sequences :math:`i_1^\ast~i_2^\ast`.
```
```suggestion
7. Let :math:`i^\ast` be the concatenation of the two sequences :math:`i_1^\ast~i_2^\ast`.
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -293,65 +293,64 @@ SIMD instructions are defined in terms of generic numeric operators applied lane
+7. Let :math:`i` be the concatenation of the two sequences :math:`i_1^\ast~i_2^\ast`.
 
-8. Let :math:`c` be the result of :math:`\lanes^{-1}_{i8x16}(i_3^{32}[\laneidx_0] \dots i_3^{32}[\laneidx_{15}])`.
+8. Let :math:`c` be the result of :math:`\lanes^{-1}_{i8x16}(i[x^\ast[0]] \dots i[x^\ast[15]])`.
```
```suggestion
8. Let :math:`c` be the result of :math:`\lanes^{-1}_{i8x16}(i^\ast[x^\ast[0]] \dots i^\ast[x^\ast[15]])`.
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -293,65 +293,64 @@ SIMD instructions are defined in terms of generic numeric operators applied lane
-      (\iff & i_1^{16} = \lanes_{i8x16}(c_1) \\
-      \wedge & i_2^{16} = \lanes_{i8x16}(c_2) \\
-      \wedge & c = \lanes^{-1}_{i8x16}((i_1^{16}~i_2^{16})[\laneidx_0] \dots (i_1^{16}~i_2^{16})[\laneidx_{15}])
+      (\iff & i = \lanes_{i8x16}(c_1)~\lanes_{i8x16}(c_2) \\
+      \wedge & c = \lanes^{-1}_{i8x16}(i[x^\ast[0]] \dots i[x^\ast[15]])
```
```suggestion
      (\iff & i^\ast = \lanes_{i8x16}(c_1)~\lanes_{i8x16}(c_2) \\
      \wedge & c = \lanes^{-1}_{i8x16}(i^\ast[x^\ast[0]] \dots i^\ast[x^\ast[15]])
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -382,74 +381,70 @@ SIMD instructions are defined in terms of generic numeric operators applied lane
 
-:math:`t_1\K{x}N\K{.}\EXTRACTLANE\K{\_}\sx^?~\laneidx`
+:math:`t_1\K{x}N\K{.}\EXTRACTLANE\K{\_}\sx^?~x`
 ......................................................
```
Nit: adjust underline

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -382,74 +381,70 @@ SIMD instructions are defined in terms of generic numeric operators applied lane
 
-:math:`\shape\K{.}\REPLACELANE~\laneidx`
+:math:`\shape\K{.}\REPLACELANE~x`
 ........................................
```
Nit: underline

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,519 @@ Where the underlying operators are non-deterministic, because they may return on
+
+
+.. _exec-vitestop:
+.. _exec-simd-alltrue:
```
Nit: just noticed that this label is named inconsistently as well.
```suggestion
.. _exec-simd-all_true:
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -182,6 +182,519 @@ Where the underlying operators are non-deterministic, because they may return on
+   \end{array}
+
+
+.. _exec-simd-anytrue:
```
```suggestion
.. _exec-simd-any_true:
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -613,19 +613,21 @@ SIMD instructions are defined in terms of generic numeric operators applied lane
 .....................................................
 
-1. Assert: due to :ref:`validation <valid-vitestop>`, two values of :ref:`value type <syntax-valtype>` |V128| are on the top of the stack.
+1. Assert: due to :ref:`syntax <syntax-instr-simd>`, that :math:`N = 2M`.
```
```suggestion
1. Assert: due to :ref:`syntax <syntax-instr-simd>`, :math:`N = 2\cdot M`.
```

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -464,6 +985,126 @@ Memory Instructions
+
+12. Let :math:`m_k` be the integer for which :math:`\bytes_{\iM}(m_k) = b^\ast[k*M/8 \slice M/8]`.
+
+13. Let :math:`J` be the integer :math:`M*2`.
```
```suggestion
13. Let :math:`J` be the integer :math:`M \cdot 2`.
```
Also, for no strong reason, I think renaming to W would look slightly nicer in the opcode below.

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -464,6 +979,125 @@ Memory Instructions
+     (\iff & \X{ea} = i + \memarg.\OFFSET \\
+     \wedge & \X{ea} + M \cdot L / 8 \leq |S.\SMEMS[F.\AMODULE.\MIMEMS[0]].\MIDATA| \\
+     \wedge & \bytes_{\iM}(m_k) = S.\SMEMS[F.\AMODULE.\MIMEMS[0]].\MIDATA[\X{ea} + k * (M/8) \slice M/8]) \\
+     \wedge & c = \lanes^{-1}_{\X{i}M\cdot2\K{x}L}(\extend^{\sx}_{M,M\cdot2}(m_0) \dots \extend^{\sx}_{M,M\cdot2}(m_{L-1}))
```
Actually, I'm sorry, I changed my mind on inlining M*2 here, since it's too obscure in the opcode.
```suggestion
     \wedge & W = M\cdot2 \\
     \wedge & c = \lanes^{-1}_{\X{i}W\K{x}L}(\extend^{\sx}_{M,W}(m_0) \dots \extend^{\sx}_{M,W}(m_{L-1}))
```

ngzhian commented:
As a checklist, the remaining open comments are (listing here since some of the comments are hidden):

- [ ] argument to narrow_u being signed https://github.com/WebAssembly/simd/pull/340#discussion_r533849824
- [ ] narrow v.s. inarrow https://github.com/WebAssembly/simd/pull/340#discussion_r534696756

rossberg commented on document/core/exec/instructions.rst:
```diff
@@ -1000,41 +1000,40 @@ Memory Instructions
-     \wedge & \bytes_{\iM}(m_k) = S.\SMEMS[F.\AMODULE.\MIMEMS[0]].\MIDATA[\X{ea} + k * (M/8) \slice M/8]) \\
-     \wedge & c = \lanes^{-1}_{\X{i}M\cdot2\K{x}L}(\extend^{\sx}_{M,M\cdot2}(m_0) \dots \extend^{\sx}_{M,M\cdot2}(m_{L-1}))
+     \wedge & \X{ea} + M \cdot N / 8 \leq |S.\SMEMS[F.\AMODULE.\MIMEMS[0]].\MIDATA| \\
+     \wedge & \bytes_{\iM}(m_k) = S.\SMEMS[F.\AMODULE.\MIMEMS[0]].\MIDATA[\X{ea} + k \cdot (M/8) \slice M/8]) \\
```
```suggestion
     \wedge & \bytes_{\iM}(m_k) = S.\SMEMS[F.\AMODULE.\MIMEMS[0]].\MIDATA[\X{ea} + k \cdot M/8 \slice M/8]) \\
```

rossberg reviewed (approved):
Great! Thanks for bearing with me.

rossberg commented:
Oops, I just noticed that there's a related issue with vtrunc and vconvert, whose execution does not map to that of a regular op either, because it involves two shapes. I think you'll need to introduce a vcvtop category and handle that separately. (Also, they are currently included in vbinop, which does not even seem correct for typing).

Feel free to fix that in a separate PR, if you prefer.

ngzhian commented:
>  I think you'll need to introduce a vcvtop category and handle that separately.

Good catch, thanks. Added a new production, `vcvtop`, which has its own validation and execution. 

The way I'm writing them now relies on this new note on execution/numerics: https://github.com/WebAssembly/simd/pull/340/files#diff-14d634364829defe70ba02a1ed0075266fce32b208b681eb5a812e96104e2bf9R41
My intention is that this also applies to the conversion operators, is that a correct assumption? Or should I clarify that section with an example showing how conversion operators work when given a list of inputs?

rossberg commented on document/core/syntax/instructions.rst:
```diff
@@ -376,7 +376,8 @@ Occasionally, it is convenient to group operators together according to the foll
+     \AVGR\K{\_u} \\
+   \production{conversion operator} & \vcvtop &::=&
      \VTRUNC\K{\_sat\_}\shape\K{\_}\sx ~|~
      \VCONVERT\K{\_}\shape\K{\_}\sx \\
```
The way you are using vcvtop in the rule patterns, you now have to exclude the shape/sx suffixes here.

ngzhian replied:
Oh yes, fixed.

rossberg reviewed (approved):
Thanks. LGTM modulo comment.

Yes, the note should suffice. If you want to emphasise that it applies to heterogeneously typed ops like conversions as well, you could replace the simple fabs example that follows with convert.
